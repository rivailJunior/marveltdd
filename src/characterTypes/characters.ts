export type CharacterDataContainer = {
    offset: number
    limit: number
    total: number
    count: number
    results: [Character]
}

export type Character = {
    id: number
    name: string
    description: string
    modified: Date
    resourceURI: String
    urls?: string[] | null
    thumbnail?: Image
    comics?: ComicList[] | any
    stories?: StoryList[] | any
    events?: EventList[] | any
    series?: SeriesList[] | any
}

export type Image = {
    path: string
    extension: string
}

export type ComicList = {
    available: number
    returned: number
    collectionURI: string
    items: [ComicSummary]
}
export type ComicSummary = {
    resourceURI: string
    name: string
}

export type StoryList = {
    available: number
    returned: number
    collectionURI: string
    items: [StorySummary]
}

export type StorySummary = {
    resourceURI: string
    name: string
    type: string
}

export type EventList = {
    items?: EventSummary[] | null
    available: number
    returned: number
    collectionURI: string
}

export type EventSummary = {
    resourceURI: string
    name: string
}

export type SeriesList = {
    items?: SeriesSummary[] | null
    available: number
    returned: number
    collectionURI: string
}

export type SeriesSummary = {
    resourceURI: string
    name: string
}