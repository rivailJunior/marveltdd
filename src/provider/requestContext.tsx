import { createContext, useContext } from "react";
import {
    Character,
    CharacterDataContainer,
} from "../characterTypes/characters";
import {
    getCharacters,
    getCharacter,
    getParticipation,
    getCharacterByName,
} from "./service";

type RequestContextType = {
    getCharacters(total: number): Promise<CharacterDataContainer>;
    getCharacter(id: number): Promise<[Character]>;
    getCharacterByName(name: string): Promise<[Character]>;
    getParticipation(id: number, type: string): Promise<any>;
};

const RequestContext = createContext({
    getCharacter,
    getCharacters,
    getParticipation,
    getCharacterByName,
});

export const useRequestContext = () => {
    const context = (useContext(
        RequestContext
    ) as unknown) as RequestContextType;
    return context;
};

export const RequestProvider = (props: any) => {
    // const getRequest = useMemo(() => {
    //     return { getCharacters, getCharacter };
    // }, []);
    // const {
    //     getParticipation,
    //     getCharacter,
    //     getCharacters,
    // } = useRequestContext();
    return (
        <RequestContext.Provider
            value={{
                getParticipation,
                getCharacter,
                getCharacters,
                getCharacterByName,
            }}
            {...props}
        />
    );
};

export default RequestContext;
