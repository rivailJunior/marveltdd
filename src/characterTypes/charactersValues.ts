
import { Character, CharacterDataContainer, ComicSummary, ComicList, StorySummary, StoryList, EventSummary, EventList, SeriesSummary, SeriesList, Image } from './characters';

export const ComicSummaryValue: ComicSummary = {
    resourceURI: "",
    name: ""
}

export const ImageValue: Image = {
    path: '',
    extension: ''
}

export const ComicListValue: ComicList = {
    available: 0,
    returned: 0,
    collectionURI: "",
    items: [ComicSummaryValue]
}

export const StorySummaryValue: StorySummary = {
    resourceURI: "",
    name: "",
    type: "",
}

export const StoryListValue: StoryList = {
    available: 0,
    returned: 0,
    collectionURI: "",
    items: [StorySummaryValue]
}

export const EventSummaryValue: EventSummary = {
    resourceURI: "",
    name: "",
}

export const EventListValue: EventList = {
    items: [EventSummaryValue],
    available: 0,
    returned: 0,
    collectionURI: "",
}

export const SeriesSummaryValue: SeriesSummary = {
    resourceURI: "",
    name: ""
}

export const SeriesListValue: SeriesList = {
    items: [SeriesSummaryValue],
    available: 0,
    returned: 0,
    collectionURI: ""

}

export const CharacterValue: Character = {
    id: 0,
    name: '',
    description: '',
    modified: new Date(),
    resourceURI: '',
    urls: [''],
    thumbnail: ImageValue,
    comics: [ComicListValue],
    stories: [StoryListValue],
    events: [EventListValue],
    series: [SeriesListValue]
}


export const CharacterDataContainerValue: CharacterDataContainer = {
    offset: 0,
    limit: 0,
    total: 0,
    count: 0,
    results: [CharacterValue]
}

