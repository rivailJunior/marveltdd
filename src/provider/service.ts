import { Character, CharacterDataContainer } from './../characterTypes/characters';
import axios, { AxiosResponse } from 'axios'
const apikey = process.env.REACT_APP_PUBKEY

export enum ParticipationTypes {
    series = "series",
    events = "events",
    comics = "comics",
}

const api = axios.create({
    baseURL: 'https://gateway.marvel.com/v1/public',
});

const getService = async (url: string, options: object): Promise<AxiosResponse> => {
    return await api.get(url, { params: { apikey, ...options } })
}

const getCharacters = async (total: number, offset?: number): Promise<CharacterDataContainer> => {
    const currentPage = offset || 0;
    const result = await getService('/characters', { offset: currentPage, limit: total })
    return result.data?.data
}

const getParticipation = async (id: number, type: string): Promise<any> => {
    const result = await getService(`/characters/${ id }/${ type }`, {})
    return result.data?.data
}

const getCharacter = async (id: number): Promise<[Character]> => {
    const result = await getService(`/characters/${ id }`, {})
    return result.data?.data?.results
}

const getCharacterByName = async (name: string): Promise<CharacterDataContainer> => {
    const result = await getService(`/characters?nameStartsWith=${ name }`, {})
    return result.data?.data
}

export { getCharacters, getCharacter, getParticipation, getCharacterByName };