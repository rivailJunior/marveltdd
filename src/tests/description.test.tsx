import React from "react";
import {
    cleanup,
    fireEvent,
    screen,
    render,
    act,
} from "@testing-library/react";
import renderer from "react-test-renderer";
import Description from "../pages/description";
import { RequestProvider } from "../provider/requestContext";
import { Router, Route } from "react-router-dom";
import { createMemoryHistory } from "history";
import { responseValueMock } from "../characterTypes/requestValueMock";

afterEach(cleanup);

const renderWithRouter = (history: any, component: any) => {
    return <Router history={history}>{component}</Router>;
};

const stateItem = responseValueMock.data.results[0];

describe("`<Description>`", () => {
    test("Render correctly", () => {
        const history = createMemoryHistory();
        history.push("/description", { ...stateItem });
        const container = renderer.create(
            renderWithRouter(history, <Description />)
        );
        container.toJSON();
        expect(container).toMatchSnapshot();
    });

    test("Show character information", () => {
        const history = createMemoryHistory();
        history.push("/description", { ...stateItem });
        act(() => {
            const { getByText, findByTestId } = render(
                renderWithRouter(history, <Description />)
            );
            expect(getByText(/Descricao do personagem/i)).toBeInTheDocument();
            expect(getByText(/Aparições/i)).toBeInTheDocument();
            // const carousel = await findByTestId("carouselItem");
            // expect(carousel.length).toBeGreaterThan(1);
        });
    });
});
