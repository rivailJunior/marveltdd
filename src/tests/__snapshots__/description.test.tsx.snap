// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<Description>\` Render correctly 1`] = `
<div
  className="jumbotron"
>
  <div
    className="container"
  >
    <h1
      className="labelHeader"
      data-testid="labelheader"
    >
      Descricao do personagem
    </h1>
    <div
      className="row"
    >
      <div
        className="col-md-6 col-12"
      >
        <div>
          <img
            alt=""
            className="picture"
            src="http://i.annihil.us/u/prod/marvel/i/mg/c/e0/535fecbbb9784.jpg"
          />
        </div>
      </div>
      <div
        className="col-md-6 col-12"
      >
        <div
          className="card"
        >
          <div
            className="card-body"
          >
            <div
              className="labelHeader card-title h5"
            >
              3-D Man
            </div>
            <div
              className="mb-2 labelHeader text-muted card-subtitle h6"
            >
              
            </div>
            <div
              className="mt-2 labelHeader text-muted card-subtitle h6"
            >
              Aparições:
            </div>
            <div
              className="labelHeader "
              style={
                Object {
                  "fontSize": 15,
                  "fontWeight": 500,
                }
              }
            >
              <ul>
                <li>
                  Quadrinhos:
                   
                  12
                </li>
                <li>
                  Eventos:
                   
                  1
                </li>
                <li>
                  Séries:
                   
                  3
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="pt-3 row"
    >
      <div
        className="col-md-4"
      >
        <div>
          Ops. Nenhuma 
          series
           encontrado
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div>
          Ops. Nenhuma 
          comics
           encontrado
        </div>
      </div>
      <div
        className="col-md-4"
      >
        <div>
          Ops. Nenhuma 
          events
           encontrado
        </div>
      </div>
    </div>
  </div>
</div>
`;
